petab/v2/core.py

```diff
@@ -0,0 +1,711 @@
+)
+
+
+class InitializationPriorType(str, Enum):
+    """Initialization prior types.
+
+    Initialization prior types as used in the PEtab parameters table.
+    """
+
+    NORMAL = C.NORMAL
+    LAPLACE = C.LAPLACE
+    UNIFORM = C.UNIFORM
+    LOG_NORMAL = C.LOG_NORMAL
+    LOG_LAPLACE = C.LOG_LAPLACE
+    PARAMETER_SCALE_NORMAL = C.PARAMETER_SCALE_NORMAL
+    PARAMETER_SCALE_LAPLACE = C.PARAMETER_SCALE_LAPLACE
+    PARAMETER_SCALE_UNIFORM = C.PARAMETER_SCALE_UNIFORM
+
+
+class Observable(BaseModel):
+    """Observable definition."""
+
+    id: str = Field(alias=C.OBSERVABLE_ID)
+    name: str | None = Field(alias=C.OBSERVABLE_NAME, default=None)
+    formula: sp.Basic | None = Field(alias=C.OBSERVABLE_FORMULA, default=None)
+    transformation: ObservableTransformation = Field(
+        alias=C.OBSERVABLE_TRANSFORMATION, default=ObservableTransformation.LIN
+    )
+    noise_formula: sp.Basic | None = Field(alias=C.NOISE_FORMULA, default=None)
+    noise_distribution: NoiseDistribution = Field(
+        alias=C.NOISE_DISTRIBUTION, default=NoiseDistribution.NORMAL
+    )
+
+    @field_validator("id")
+    @classmethod
+    def validate_id(cls, v):
+        if not v:
+            raise ValueError("ID must not be empty.")
+        if not is_valid_identifier(v):
+            raise ValueError(f"Invalid ID: {v}")
+        return v
+
+    @field_validator(
+        "name",
+        "formula",
+        "noise_formula",
+        "noise_formula",
+        "noise_distribution",
+        "transformation",
+        mode="before",
+    )
+    @classmethod
+    def convert_nan_to_none(cls, v, info: ValidationInfo):
+        if isinstance(v, float) and np.isnan(v):
+            return cls.model_fields[info.field_name].default
+        return v
+
+    @field_validator("formula", "noise_formula", mode="before")
+    @classmethod
+    def sympify(cls, v):
+        if v is None or isinstance(v, sp.Basic):
+            return v
+        if isinstance(v, float) and np.isnan(v):
+            return None
+
+        return sympify_petab(v)
+
+    class Config:
+        populate_by_name = True
+        arbitrary_types_allowed = True
+
+
+class ObservablesTable(BaseModel):
+    """PEtab observables table."""
+
+    observables: list[Observable]
+
+    def __getitem__(self, observable_id: str) -> Observable:
+        """Get an observable by ID."""
+        for observable in self.observables:
+            if observable.id == observable_id:
+                return observable
+        raise KeyError(f"Observable ID {observable_id} not found")
+
+    @classmethod
+    def from_dataframe(cls, df: pd.DataFrame) -> ObservablesTable:
+        if df is None:
+            return cls(observables=[])
+
+        observables = [
+            Observable(**row.to_dict())
+            for _, row in df.reset_index().iterrows()
+        ]
+
+        return cls(observables=observables)
+
+    def to_dataframe(self) -> pd.DataFrame:
```

Comment:
Fine as is, but since df is common enough
```suggestion
    def from_df(cls, df: pd.DataFrame) -> ObservablesTable:
        if df is None:
            return cls(observables=[])

        observables = [
            Observable(**row.to_dict())
            for _, row in df.reset_index().iterrows()
        ]

        return cls(observables=observables)

    def to_df(self) -> pd.DataFrame:
```